### DISCUSSION TITLE
Thermal runaway modelling

### DISCUSSION THREAD

**Discussion by ahsadeghi** [pybamm-team/PyBaMM#3026] [Q&A]
Hello,
Is it possible to model thermal runaway in Pybamm?

Hosein

**Comment by brosaplanella** [comment]
We have thermal models, but I don't think they can capture thermal runaway. In fact, I think it is still an open question how to model thermal runaway.

**Comment by ahsadeghi** [comment]
Hi Ferran @brosaplanella,
I want to try the implementation of thermal runaway models in Pybamm. Could you please tell me where are the energy conservation equations in the code? and do I need to work with the developer version or can I do it in the input file, similar to the modification of model parameters (the workshop in Oxford)?

Regards,
Hosein

**Reply by brosaplanella** [reply]
Hi Hosein! You can do it with PyBaMM as a dependency only (rather than a develop install). The steps would be the following:
1. Create a repo using the [PyBaMM cookiecutter](https://github.com/pybamm-team/pybamm-cookie)
2. Add the new model in the `src/.../models/input` directory. You can use the [Basic DFN model](https://github.com/pybamm-team/PyBaMM/blob/develop/src/pybamm/models/full_battery_models/lithium_ion/basic_dfn.py) as a starting point.
3. Add the new physics to it. We don't have an example of DFN + thermal, but you can check the [lumped submodel](https://github.com/pybamm-team/PyBaMM/blob/develop/src/pybamm/models/submodels/thermal/lumped.py) (which inherits from the [base thermal submodel](https://github.com/pybamm-team/PyBaMM/blob/develop/src/pybamm/models/submodels/thermal/base_thermal.py)) to see how the energy balance is implemented in PyBaMM.

**Reply by ahsadeghi** [reply]
Hello @santacodes 
Sorry, I have problems with step 3. I have added the lumped model to the DFN model, then I called the thermal model as an option (`options = {"thermal": "lumped"}`) but it is not recognised when running. Could you please explain the procedure for step 3?

Regards,
Hosein

**Comment by ahsadeghi** [comment]
Thank you. I followed the steps, now I have a project folder (with the default name pybamm_example_project) containing the following items
![image](https://github.com/user-attachments/assets/dfcadb1b-6c1b-404f-917b-2e2ded62af79)

Then I copied `basic_dfn.py` to `...\pybamm-example project\src\pybamm_example_project\models\input` and now I have the following items there.
![image](https://github.com/user-attachments/assets/bb5f3b28-30e8-4abe-bcf3-8472ec0e2297)

Is it correct so far?
Then when I add the physics to `basic_dfn.py`, do I need to compile the code or can I modify the DFN model in the input file and run it with my local PyBaMM?

Regards,
Hosein

**Reply by brosaplanella** [reply]
Hi, yes! When you create the repo it asks you a few questions, so you can set your own project name (I would actually advise to create a new project and give it a more meaningful name, such as thermal-runaway, as this name will be used quite frequently). You can modify the basic_dfn.py directly (you can also rename it). In order to use it, you will need to add an extra line in the `pyproject.toml` under the model entry points section (you can just copy one of the existing lines and change it to point to your new model). Then, from the main script, you can call your model as `your_project_name.Model("your model name")`.

**Reply by agriyakhetarpal** [reply]
Also, pinging @santacodes here, the project author, who can potentially act as an additional resource and help answer any other questions related to usage and setup.

**Comment by ahsadeghi** [comment]
Hi, thank you. So now I have the following lines in `pyproject.toml` 

`[project.entry-points."models"]`
`SPM = "thermal_runaway.models.input.SPM:SPM" `
`BasicReservoir = "thermal_runaway.models.input.BasicReservoir:BasicReservoir"`
`DFN= "thermal_runaway.models.input.basic_dfn:basic_dfn"`

Then what is the main script here? do you mean my (arbitrary) simulation script?

**Reply by santacodes** [reply]
Hey I'm not sure why you have put `thermal_runaway` in the entry point, I think you mentioned earlier that you set up the project with defaults, and mentioned the project name was `pybamm_example_project`. But if you actually did set it up with the project name `thermal_runaway` then this should work after you install this package locally by doing something like `pip install -e .` while being inside the project directory in the command prompt. Once the pip installation is done you can then perform `import pybamm_example_project` (given your project name is pybamm_example_project) inside a python file and then load a model by using the `Model()` method, which would look something like `spm_model = pybamm_example_project.Model("SPM")`. 
This would load the spm model object to the variable spm_model.

**Reply by ahsadeghi** [reply]
yes, I changed the name to `thermal_runaway` as Ferran suggested.
I did the command `pip install e .` and now it says that `thermal_runaway `package is installed.
What I want to do is calculating and adding a source term to the energy conservation equation for including the thermal runaway heat generation and the independent variable is cell temperature. Isn't it better that I only work on the [lumped submodel ](https://github.com/pybamm-team/PyBaMM/blob/develop/src/pybamm/models/submodels/thermal/lumped.py) instead of modifying the DFN model?

**Comment by ahsadeghi** [comment]
Hi @santacodes 
After installing `thermal_runaway`, I am trying to import it in an input script, as follows

```
import pybamm
import thermal_runaway
model = thermal_runaway.Model("DFN")
sim = pybamm.Simulation(model)
sim.solve([0, 3600])
sim.plot()
```

but I get the following error 

```
Traceback (most recent call last):
  File "C:\Users\k21140758\OneDrive - King's College London\Documents\7-pybamm_thermal_runaway\tut_01.py", line 5, in <module>
    model = thermal_runaway.Model("DFN")
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\k21140758\OneDrive - King's College London\Documents\7-pybamm_thermal_runaway\thermal_runaway\src\thermal_runaway\entry_point.py", line 154, in Model
    return models[model]
           ~~~~~~^^^^^^^
  File "C:\Users\k21140758\OneDrive - King's College London\Documents\7-pybamm_thermal_runaway\thermal_runaway\src\thermal_runaway\entry_point.py", line 97, in __getitem__
    return self._load_entry_point(key)()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\k21140758\OneDrive - King's College London\Documents\7-pybamm_thermal_runaway\thermal_runaway\src\thermal_runaway\entry_point.py", line 106, in _load_entry_point
    ps = self._all_entries[key] = ps.load()
                                  ^^^^^^^^^
  File "C:\Users\k21140758\AppData\Local\Programs\Python\Python311\Lib\importlib\metadata\__init__.py", line 198, in load
    module = import_module(match.group('module'))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\k21140758\AppData\Local\Programs\Python\Python311\Lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\Users\k21140758\OneDrive - King's College London\Documents\7-pybamm_thermal_runaway\thermal_runaway\src\thermal_runaway\models\input\DFN.py", line 5, in <module>
    from .base_lithium_ion_model import BaseModel
ModuleNotFoundError: No module named 'thermal_runaway.models.input.base_lithium_ion_model'
```

But it works with SPM model

`model = thermal_runaway.Model("SPM")
`

I also added the DFN model in pyproject.toml before installaiton

```
[project.entry-points."models"]
SPM = "thermal_runaway.models.input.SPM:SPM"
BasicReservoir = "thermal_runaway.models.input.BasicReservoir:BasicReservoir"
DFN= "thermal_runaway.models.input.DFN:DFN"
```

Could you please help me to fix the error.

Regards,
Hosein

**Reply by santacodes** [reply]
Hey I'm guessing this is giving an error because the `BaseModel` is not contained within your generated pybamm-cookie project. You could fix this by changing the line in dfn.py where it imports the BaseModel to `import pybamm` and then inside the dfn class inherit the BaseModel using something like this, `class DFN(pybamm.lithium_ion.BaseModel)`, this should work, if it doesn't you could refer to the `spm.py` in the models/input folder. It has the same implementation. If it still gives an error feel free to post it, sorry if there are delays between my messages because of timezones :).

**Reply by brosaplanella** [reply]
Yes, I think the DFN should inherit from `pybamm.lithium_ion.BaseModel` (the example file has a relative import, hence the error).

**Reply by ahsadeghi** [reply]
it worked, thank you!

**Comment by ahsadeghi** [comment]
Dear Ferran,
I am trying to implement the thermal model in the DFN model but it has not worked yet.
After several modifications I am getting the following error:
`KeyError: "'Volume-averaged cell temperature [K]' not found. Best matches are []"`
Please see the files attached. Could you please help me with debugging it, or could you please show me a close example of such implementation.

[base_thermal.txt](https://github.com/user-attachments/files/18039892/base_thermal.txt)
[DFN.txt](https://github.com/user-attachments/files/18039893/DFN.txt)
[lumped.txt](https://github.com/user-attachments/files/18039894/lumped.txt)

Regards,
Hosein

**Reply by brosaplanella** [reply]
Hi! Just to clarify, which model are you calling to get this error? The error says that the variable is not defined, which means whatever model you are running will have to include it. It would also be helpful to get a minimum working example to see which line of code is triggering the error.

**Comment by ParthSwaroop** [comment]
I have implemented a 2-eq model thermal runaway model using on solve_ivp and events function in Python based on this literature framework : Chen, Haodong, et al. "A simplified mathematical model for heating-induced thermal runaway of lithium-ion batteries." Journal of The Electrochemical Society 168.1 (2021): 010502.
I also was able to do a parametric runs for Cp, H and reaction parameters ( Ai, Ei, Hi) of the equations mentioned in the literature.  The solver is also validated with our in-house cell data.
Do you think it would be helpful for you guys if i'm able to convert this code into PyBAMM setup ?  My codes are for self use only and I have never worked on github before :) I was just browsing through your work and decided to test the cell thermal performance and aging models.
I'm primarily a flow thermal engineer, I use python only for post processing and sometimes for statistical analysis

**Reply by brosaplanella** [reply]
Hi! The proposed way to share models now is to build them as standalone models and publish them in your own Github repository (under the license of your choice). You can then share with the community by posting them in our Discourse ["Model sharing" section](https://pybamm.discourse.group/c/model-sharing/5). 

We advise to use [this template](https://github.com/pybamm-team/pybamm-cookie) to create the repo, which provides the necessary setup.
